# 2025. 02. 26.

## Reactor 3 Reference Guide

### Reactor Core 기능

#### 오류 처리

##### 1. 오류 처리 연산자

###### 1.6. 로깅 또는 따로 반응

오류를 계속 전파하면서 시퀀스를 변형하지 않고 반응(예를 들어 로깅)하고자 하는 경우, `doOnError` 연산자를 사용할 수 있다. 이는 다음과 같은 "오류를 잡아 오류 메시지를 로그로 남긴 다음 다시 던지"는 패턴과 동등하다:

```java
try {
  return callExternalService(k);
}
catch (RuntimeException error) {
  //make a record of the error
  log("uh oh, falling back, service failed for key " + k);
  throw error;
}
```

`doOnError` 연산자 뿐만아니라 `doOn` 접두어를 가진 모든 연산자가 때로는 "부수 효과"를 갖는 것처럼 불린다. 시퀀스의 이벤트를 변형하지 않고 슬쩍 볼(peek) 있게 한다.

위의 절차적 예시와 마찬가지로 다음 예시는 여전히 오류를 전파하지만 외부 서비스가 실패했음을 확인하는 로그를 남긴다:

```java
LongAdder failureStat = new LongAdder();
Flux<String> flux =
Flux.just("unknown")
    .flatMap(k -> callExternalService(k) // 1. 실패할 수 있는 외부 서비스 호출
        .doOnError(e -> {
            failureStat.increment();
            log("uh oh, falling back, service failed for key " + k); // 2. ..을 로깅과 통계 사이드이펙트로 꾸민(decorate) 다음
        }) 
        // 3. 이곳에 오류 복구 연산자를 사용하지 않는 한 오류로 끝난다.
    );
```

또한 통계 카운터를 증가시키는 것을 두 번째 부수 효과로 볼 수도 있다.